[2단계 - 페이먼츠] 블루(이한희) 미션 제출합니다.

hanheel opened the pull request:
## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [ ] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [ ] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입
    - 페이먼츠 웹 :  https://hanheel.github.io/react-payments/
    - 스토리북 : https://680d9229aa540cfc08d235a8-vunbroirvu.chromatic.com/?path=/docs/components-card--docs

## ✅ 기능 요구사항

### 동적 입력 UI 구현

- 입력 필드는 사용자의 입력이 완료되면 다음 필드로 자동으로 이동한다.

### 카드사 선택

- 사용자는 카드사를 선택할 수 있고, 카드사에 따라 미리보기 카드의 색상을 변경한다.

### CVC 번호

- 입력은 숫자만 가능하며, 유효하지 않은 값을 입력 시 피드백을 제공한다.

### 폼 제출 및 상태 관리

- 모든 카드 정보가 정확하게 입력되고 검증되었을 때 확인 버튼이 활성화된다.
- 사용자가 입력한 정보 중 하나라도 지우거나 수정하여 유효하지 않게 되면, 확인 버튼은 보이지 않는다.

### 카드 등록 완료 및 네비게이션

- 확인 버튼을 클릭하면 사용자는 카드 등록 완료 페이지로 이동한다.
- 카드 등록 완료 페이지에서는 카드 등록이 성공적으로 완료되었다는 메시지와 함께, 다시 카드 등록 페이지로 돌아갈 수 있는 확인 버튼을 제공한다.

### 스토리

- 카드사에 따라 미리보기 카드의 색상이 변경된 스토리 제작
- Input 종류별로 기본 렌더링 상태와 에러 상태를 반영한 스토리 제작

## 💬 의도

### 1️⃣ 컴포넌트 통합

- step1에서의 구조 : `cardNumberInput` , `cvcInputs` , `expirationDateInput` 이 따로 존재
- step2에서의 구조 : `CardLabeledInput` 컴포넌트로 통일

step1에서는 cardNumberInput, cvcInputs, expirationDateInput처럼 입력 컴포넌트가 각각 분리되어 있었습니다. 그러나 step2로 넘어오면서 카드 브랜드, 카드 비밀번호 등 새로운 입력 항목이 추가되면서, 코드 중복이 많아지고 하나의 폼에서 관리해야 할 부분이 크게 늘어났습니다.
이러한 문제를 해결하기 위해 CardLabeledInput이라는 하나의 컴포넌트로 모든 입력을 통합하여 일관된 구조로 관리할 수 있도록 개선하였습니다.

### 2️⃣ UI config 분리 (`CardLabeledInputConfig` )

1번의 컴포넌트 통합을 진행하면서, props의 수가 자연스럽게 늘어나는 문제가 발생했습니다.  `id`, `label`, `inputKeys`, `placeholder`, `maxLength` 등 각 입력 항목에 필요한 속성들을 관리하기 위해 별도의 설정 객체를 만들어, `config` 라는 이름으로 props에 전달하였습니다

```tsx
  cardNumber: {
    id: 'card-number',
    label: '카드 번호',
    inputKeys: ['first', 'second', 'third', 'fourth'],
    placeholder: '1234',
    maxLength: 4,
  },
  cardBrand: {
    id: 'card-brand',
    label: '카드브랜드',
    options: CARD_BRANDS,
    defaultMessage: '카드사를 선택해 주세요',
  },
```

## 💬 질문

### ❓ 컴포넌트의 계층 구조

현재 컴포넌트 구조는 다음과 같습니다.

```tsx

CardDescriptiveInput → CardLabeledInput → CardInput

```

- **CardInput**: 가장 기본적인 입력 필드 (재사용 가능)
- **CardLabeledInput**: 라벨과 에러 메시지를 포함한 입력 그룹
- **CardDescriptiveInput**: 설명 텍스트를 함께 보여주는 입력 컴포넌트 컨테이너

step1에서는 입력 종류별로 서로 다른 컴포넌트로 관리했지만, 관심사 분리 측면에서 컴포넌트를 재설계해보았습니다.

그 결과 구조적으로 겹치는 부분을 줄이고, 공통적으로 관리해야 할 부분을 쉽게 파악할 수 있게 되어 유지보수성이 개선되었다고 생각했습니다.

다만 이전에 비해 전달해야 할 props의 수가 많아졌고, 이로 인해 step1에 비해 오히려 복잡성이 증가한 것은 아닌지 고민이 됩니다.

현재로서는 props drilling이 아주 깊진 않아 Context API 도입까지는 고려하지 않았습니다.

1. 혹시 **현재 코드 구조**에서 복잡성을 더 줄이기 위해 추가로 시도해볼 수 있는 방법이 있을지 궁금합니다.
2. (의도 2번) config를 통해 일부의 props를 전달하는 방식이 코드를 읽을 때 바로 props를 파악하기 어렵게 만들거나 컴포넌트 간 연결 관계를 추적하는데에 오히려 복잡성을 증가시킨건 아닌가 하는 고민이 듭니다. 현재 수준의 웹에서 위와같은 방식 외에 시도해 볼 수 있는 다른 방식이 있을까요? (config를 사용한 방식 외에, 복잡성을 줄이기 위해 제가 시도해 볼 수 있는 방식이 있을까요?)

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

hanheel committed:
docs : 기능 요구사항 정리

hanheel committed:
style : 디자인 시스템 정리

hanheel committed:
style : reset.css 설정

hanheel committed:
feat : Description 컴포넌트 제작

hanheel committed:
feat : Input 컴포넌트 제작

bunju20 committed:
feat: 카드 컴포넌트 생성
- 카드 엠블럼 변경 로직 추가 필요

hanheel committed:
feat : 카드 타입에 따라 visa 또는 mastercard 로고 생성

hanheel committed:
feat : LabeledInput 컴포넌트 제작

hanheel committed:
feat : 레이블 input 타입 매칭

bunju20 committed:
feat: pages UI 제작
- 스타일 추가 필요

bunju20 committed:
style: pages css gap 변경

hanheel committed:
feat : 카드 번호를 입력 받고 카드 프리뷰에 반영

hanheel committed:
fix : input maxLength 설정되지 않은 오류 해결

bunju20 committed:
fix: 카드 번호 위치에 맞도록 수정

bunju20 committed:
feat: 유효성 검사 및 오류 처리 로직 구현

hanheel committed:
fix : input값이 0인 경우 유효성 검사 배제

hanheel committed:
feat : 카드번호 마스킹 기능 제작

hanheel committed:
docs : 기능 요구사항 업데이트

bunju20 committed:
feat: 카드 첫 번호에 따라 BrandLogo 변경 로직 구현

hanheel committed:
refactor : 에러메세지 상수 분리

hanheel committed:
refactor : justifyLogoBrand 함수 분리

hanheel committed:
refactor : 불필요한 errorMessage 삭제

hanheel committed:
feat : InputGroup 스토리 작성

bunju20 committed:
feat: storyBook 배포 및 inputGroup 수정

bunju20 committed:
feat: 카드 스토리 (visa, default, mastercar)추가

bunju20 committed:
feat: Card Story (카드번호, 유효기간) 구현

hanheel committed:
refactor : 불필요한 style 삭제

hanheel committed:
fix : 에러 메세지가 덮어씌워지는 오류 해결

hanheel committed:
refactor : erroMessages에 함수대신 상수 전달

hanheel committed:
feat : 카드 번호 유효기간 실시간 입력 반영

hanheel committed:
refactor : maskingNumber 유틸함수 card 폴더 내로 이동

hanheel committed:
fix : 스토리북 에러 메세지 해결

hanheel committed:
chore : 배포 설정

hanheel committed:
chore : storybook 정적파일 깃 ignore 설정

hanheel committed:
chore : storybook 정적파일 캐시 삭제

hanheel committed:
feat : globalStyle 적용

hanheel committed:
refactor : 스토리북과 컴포넌트를 가깝게 위치

1. preview.ts에 globalStyle 설정
2. 스토리북과 컴포넌트를 가깝게 위치하도록 폴더 수정

hanheel committed:
refactor : 유틸 함수 위치 src/util로 수정

hanheel committed:
refactor : Input을 CardInput으로 변경

hanheel committed:
refactor : AddCard에서 CardNumberInput 컴포넌트 분리

hanheel committed:
refactor : AddCard에서 ExpirationDateInputs 컴포넌트 분리

hanheel committed:
refactor : AddCard에서 CVCInputs 컴포넌트 분리

hanheel committed:
refactor : div 태그를 form으로 변경

hanheel committed:
refactor : CardNumberInputs에서 반복되는 CardInput 추상화

hanheel committed:
refactor : ExpirationDateInputs에서 반복되는 CardInput 추상화

hanheel committed:
refactor : getFirstErrorMessage 추상화

hanheel committed:
refactor : styled-component의 위치를 로직 하단으로 수정

hanheel committed:
refactor: CardNumber 컴포넌트 분리

hanheel committed:
refactor : InputType을 CardInputProps의 key로 통일

hanheel committed:
feat : label과 input을 id로 연결

hanheel committed:
refactor : InputProps가 ComponentProps를 확장하도록 변경

hanheel committed:
refactor : 스토리 코드 파일 분리

hanheel committed:
refactor : 사용하지 않는 state 삭제

hanheel committed:
feat : Select 컴포넌트 및 열고 닫힘 기능 구현

hanheel committed:
feat : 배경을 클릭할 경우 select 박스가 닫힘

hanheel committed:
feat : 비밀번호 입력 input 제작

hanheel committed:
refactor : description text 상수 분리

hanheel committed:
refactor : 공통 훅 분리 및 cardInputs 재활용하는 형태로 수정

hanheel committed:
refactor : input config 상수 분리

hanheel committed:
feat : select 된 값을 저장 후 UI에 반영

hanheel committed:
refactor : 상수를 한 파일로 통일

hanheel committed:
feat : 카드 브랜드 선택 시 카드 색상 변경

hanheel committed:
feat : options 애니메이션

hanheel committed:
feat : 유효하지 않는 값인 경우 입력 제한

hanheel committed:
feat : 단계별 렌더링

hanheel committed:
feat : 등장 애니메이션 적용

hanheel committed:
feat : 확인 버튼을 누를 경우 제출 페이지로 이동

hanheel committed:
feat : 자동 Focus 기능

hanheel committed:
feat : 첫 input focus

hanheel committed:
feat : 유효기간 형식 에러 추가

hanheel committed:
refactor : todo 주석 삭제

hanheel committed:
feat : CardLabeledInput 스토리 작성

hanheel committed:
feat : CardLabeledSelect 스토리 작성

hanheel committed:
feat : 카드 브랜드 선택 스토리 작성

hanheel committed:
fix : 폼 자동제출 방지

hanheel committed:
refactor : focus 기능 하나의 책임으로 통일

hanheel committed:
refactor : 상태 전달 방식을 navigate의 state로 수정

wonsss commented:
안녕하세요 블루~!
코드 리뷰를 하면서 살펴보니, 2단계 요구사항 구현이 없는 것 같아 이상하여 살펴보니..
2단계 PR의 from branch가 stpe2가 아닌  step1으로 잘못 지정하여 작성해주셨네요.ㅠ 

<img width="882" alt="image" src="https://github.com/user-attachments/assets/95b5c7b7-ac78-46f3-a9e4-37ef9c535802" />

hanheel committed:
Merge branch 'hanheel' into step1

hanheel commented:
헉 죄송합니다!!!! 
방금 step1 브랜치에 step2의 변경사항을 머지해서 반영했습니다!
혼란을 드려 죄송합니다... 앞으로는 한 번 더 확인하는 습관을 들이도록 하겠습니다!

wonsss commented on src/routes.tsx:
```diff
@@ -0,0 +1,20 @@
+    },
+  ],
+  { basename: '/react-payments/' }
+);
```
createBrowserRouter 함수 방식으로 잘 사용해주셨네요! 👍

wonsss commented on src/routes.tsx:
```diff
@@ -0,0 +1,20 @@
+      element: <Layout />,
+      children: [
+        { path: '/', element: <AddCard /> },
+        { path: '/success', element: <Success /> },
```
AddCard 와 Success 컴포넌트명 이름만 보면 "화면(page)"인지, "일반 UI 컴포넌트"인지 직관적으로 구분이 잘 되지 않는 느낌이 있습니다.

컴포넌트 네이밍에 page의 역할을 명시적으로 드러내도록 AddCardPage, SuccessPage 처럼 네이밍하는 규칙을 고려해볼 수 있을 것 같아요.

hanheel replied:
이름에서 역할이 명확히 드러나야 웹의 크기가 커졌을 때 명확하게 "페이지"임을 인식할 수 있겠군요..! 감사합니다 :)

✅ 커밋 [4512f2e](https://github.com/woowacourse/react-payments/pull/465/commits/4512f2eaac2851e84461c5a161cf195241bf457c)

wonsss commented on src/routes.tsx:
```diff
@@ -0,0 +1,20 @@
+import Success from './pages/Success';
+import Layout from './pages/Layout';
+
+const router = createBrowserRouter(
```
참고로 createBrowserRouter 외에도 다른 라우터 방식들도 있는데, 각 어떤 경우에 사용하면 적절한지도 한번 알아보시면 좋을 것 같습니다.

[createBrowserRouter](https://reactrouter.com/6.30.0/routers/create-browser-router)
[createMemoryRouter](https://reactrouter.com/6.30.0/routers/create-memory-router)
[createHashRouter](https://reactrouter.com/6.30.0/routers/create-hash-router)
[createStaticRouter](https://reactrouter.com/6.30.0/routers/create-static-router)

wonsss commented on src/styles/FormSectionStyle.ts:
```diff
@@ -0,0 +1,10 @@
+  display: flex;
+  flex-direction: column;
+  gap: 16px;
+  animation: ${slideDown} 0.5s ease-in-out;
```
애니메이션 잘 주셨네요! 센스 👍

wonsss commented on src/types/CardInputTypes.ts:
```diff
@@ -1,9 +1,13 @@
+  YY: string;
+  CVC: string;
+  password: string;
+  cardBrand: keyof typeof cardBrandsColors;
```
유틸 타입을 잘 활용해주셨네요 👍

wonsss commented on src/validation/getCardErrorMessages.ts:
```diff
@@ -0,0 +1,66 @@
+};
+
+export const getCardExpirationYYErrorMessage = (expirationDate: string) => {
+  if (expirationDate.length === 0) return;
+  if (!isValidLengthYear(expirationDate)) return;
+
+  if (isNumber(expirationDate)) {
+    return ERROR_MESSAGE.NOT_A_NUMBER;
+  }
+  if (!isValidYearRange(expirationDate)) {
+    return ERROR_MESSAGE.INVALID_EXPIRATION_YEAR;
+  }
```
사소하긴 한데, 조건문 return이 한줄일 때 중괄호를 생략해도 되지만, 이처럼 생략하거나 생략하지 않은 경우를 혼용하기보다 일관성있게 컨벤션을 잡는 게 좋습니다.

일반적으로 조건문에서 중괄호를 사용하는 것이 더 가독성이 좋습니다. 단일 구문이더라도 중괄호를 사용하는 것이 의도를 명확히 하고, 코드 확장 시 실수를 방지하는 데 유리합니다.

hanheel replied:
협업자의 측면에서 생각해봤을 때 확실히 통일하는 것이 일관성을 지키게 돼서 가독성이 좋아지겠네요! 반영하겠습니다 감사합니다 ☺️

✅ 커밋 : fcdc67d1cd7c364069149a37d7cd31d0164f88aa

wonsss commented on src/pages/Success.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  if (!cardFirstNumber || !cardBrand) {
+    return <Message>오류가 발생했어요! 다시 시도해 주세요.</Message>;
+  }
```
예외처리도 꼼꼼히 잘 해주셨네요! 👍

wonsss commented on src/pages/AddCard.tsx:
```diff
@@ -1,94 +1,126 @@
+  const { stepIndex, handleStepIndex } = useCardFormStep({ cardInput, isError });
+
+  const handleSubmit = () => {
+    navigate('/success', {
```
path도 상수로 관리하면 유지보수 시 용이할 것 같아요.

hanheel replied:
페이지가 많아져 웹이 커졌을 때 추적이 용이해지겠네요!! 감사합니다 👍🏻

✅ 커밋 : 116b8e8

wonsss commented on src/pages/AddCard.tsx:
```diff
@@ -1,94 +1,126 @@
+
+  useEffect(() => {
+    handleStepIndex();
+  }, [cardInput, isError]);
```
현재 handleStepIndex를 useEffect에서 호출하고 있는데, 이는 불필요하게 useEffect를 사용한 방식으로 비즈니스 로직을 처리하는 것은 아닐지 고민해보면 좋을 것 같아요. React에서는 useEffect를 남용하지 않고, 동기적인 상태 변경을 처리하는 것이 더 적합하고 간결한 코드를 작성하는 데 도움이 됩니다. 
handleStepIndex는 상태 변경 직후 실행되는 것이 직관적이고 효율적일 것 같아요.
또한, useEffect는 상태가 변경될 때마다 불필요한 재렌더링을 유발할 리스크도 있습니다

예를 들어, handleCardInput 함수 내에서 setCardInput 상태 업데이트 후 즉시 handleStepIndex(updatedInput) 함수에 갱신된 상태 넣어서 호출하는 것은 어떨까요?

```ts
  const handleCardInput = (inputKey: keyof CardInputProps, value: string) => {
    if (!validateCardInputs(inputKey, value)) return;

    handleErrorMessages(inputKey, '');
    setIsError((prev) => ({
      ...prev,
      [inputKey]: false,
    }));

    setCardInput((prev: CardInputProps) => {
      const updatedInput = { ...prev, [inputKey]: value };
      handleStepIndex(updatedInput);
      return updatedInput;
    });
  };
```

참고로 React 공식 문서에서 useEffect에 대한 좋은 글이 있는데, 읽어보시길 추천드려요.

https://react.dev/learn/you-might-not-need-an-effect

hanheel replied:
좋은 아티클 공유 감사드립니다 😁

`useEffect` 가 의존성 배열을 관리하는 측면에서나 로직 분산 측면에서나 과다하게 사용하면 예상치 못하게 동작하는 부분이 많아지는 것 같아 줄여보고 싶었습니다. 

따라서 `handleStepIndex`를 `useForm` 내부의 `handleCardInput` 함수에 넣는 것도 고려했지만,  `useForm` 훅이 입력, 검증, 에러, step관리 흐름까지 모두 책임지게 되어 책임이 과도하게 커진다고 판단했습니다. 현재 useForm은 사용자의 input을 받고 에러 여부를 반환하는 "데이터"를 다루는 형태인데 step과 같은 UI 흐름제어에도 사용을 하는 것이 맞을까? 하는 의문이 들었습니다.

그래서 현재는 cardInput 변경 시점에 useEffect로 stepIndex를 계산하는 방식이 가장 구조적으로 납득 가능한 선택이라고 판단했습니다. 위와 같은 관점에서 useEffect를 사용하는 구조는 괜찮은 방향일지, 혹은 입력과 동시에 흐름 제어까지 훅 내부에서 처리하는 방식이 더 적절할지 조언을 주시면 감사드리겠습니다!!!

wonsss replied:
블루의 말씀대로 로직의 관심사 분리도 중요하며 적절한 의도로 잘 개발해주셨어요.
다만, useEffect는 비동기성/사이드 이펙트를 위해 사용되어야 하며, 단순한 상태 연쇄 로직에는 적합하지 않다고 생각해요. 
의존성 배열 관리는 복잡하고, 예기치 않은 실행을 유발할 가능성이 있어요.
setCardInput 직후 handleStepIndex(updatedInput)를 호출하면 더 직관적이고 성능 상 이점이 있어요.

wonsss commented on src/validation/validateConditions.ts:
```diff
@@ -0,0 +1,26 @@
+
+export const isNumber = (input: string) => {
+  return Number.isNaN(Number(input));
+};
```
isNumber 함수의 구현이 반대로 동작하고 있습니다.

hanheel replied:
반영했습니다!! 👍🏻

✅커밋 : [335fe72](https://github.com/woowacourse/react-payments/pull/465/commits/335fe72a9357fb1e33d436e6e7e9927d3ad13de7)

wonsss commented on src/hook/useOutsideClick.ts:
```diff
@@ -0,0 +1,25 @@
+import { useEffect } from "react";
+
+const useOutsideClick = (
+  ref: React.RefObject<HTMLElement | null>,
+  callback: () => void
+) => {
+  useEffect(() => {
+    const handleClickOutside = (e: MouseEvent) => {
+      if (
+        ref.current &&
+        e.target instanceof Node &&
+        !ref.current.contains(e.target)
+      )
+        callback();
+    };
+
+    window.addEventListener("mousedown", handleClickOutside);
+
+    return () => {
+      window.addEventListener("mousedown", handleClickOutside);
+    };
+  }, [callback, ref]);
+};
+
+export default useOutsideClick;
```
useOutsideClick은 자주 사용되는 커스텀 훅이죠. 잘 적용해주셨네요👍

wonsss commented on src/hook/useForm.ts:
```diff
@@ -0,0 +1,83 @@
+    CVC: false,
+    password: false,
+  });
+  const { errorMessages, handleErrorMessages } = useErrorMessages();
```
isError 상태가 반드시 필요한가요? 기존 errorMessages 상태를 통해 파생된 값으로 에러 여부를 판단할 수 있지 않을까요? 
예를 들어, errorMessages.first 가 비어 있지 않다면 에러가 있다고 간주하면 될 것 같아서요.
꼭 필요한 상태만 관리하고 상태 관리는 단순하게 하는 게 유지보수 및 성능에 좋아요.

```
  const isError = {
    first: errorMessages.first !== '',
    second: errorMessages.second !== '',
    third: errorMessages.third !== '',
    fourth: errorMessages.fourth !== '',
    MM: errorMessages.MM !== '',
    YY: errorMessages.YY !== '',
    CVC: errorMessages.CVC !== '',
    password: errorMessages.password !== '',
  };
```

hanheel replied:
공식 문서를 공부하다가 "연산으로 도출될 수 있는 값은 상태로 두지 않는 것이 좋다"는 내용을 봤었는데, 

step1 구조에서는 `isError`가 해당 form의 필드 전체에 대한 에러 여부를 담고 있었고, `errorMessage`는 `getFirstMessage`로 추출한 하나의 메시지만 반환하도록 되어 있어서 그 부분을 당시에 바로 반영하진 못했습니다.

이번 step2에서는 컴포넌트에 전달되는 isError와 errorMessage의 구조가 같아지면서 자연스럽게 리팩토링이 가능했던 것 같습니다. 앞으로는 props의 구조에 대해서도 고민해봐야겠네요..! 좋은 피드백 감사드립니다!! 

✅ 커밋 : [510d922](https://github.com/woowacourse/react-payments/pull/465/commits/510d922dbc28ee04202b85de3f928996d0592ce0)

wonsss replied:
공식 문서를 꼼꼼히 잘 읽으셨네요 👍

wonsss commented on src/pages/AddCard.tsx:
```diff
@@ -1,94 +1,126 @@
-        />
+        {stepIndex >= 5 && (
+          <ButtonContainer>
+            <Button text="확인" handleClick={handleSubmit} />
```
현재 onClick 이벤트를 사용하여 폼 제출을 처리하고 있습니다. 하지만 폼 제출은 브라우저 기본 동작 스펙에 맞게 onSubmit 이벤트를 활용하는 것이 일반적으로 권장됩니다. 
예를 들어, onClick 이벤트로 구현된 현재 코드에서는 form을 다 입력하고 Enter 키를 눌러도 폼이 제출되지 않기 때문에 사용성 측면에서 제한적입니다. 
onSubmit 이벤트를 사용하면 브라우저의 기본 폼 제출 동작을 활용할 수 있으며, 이는 Enter 키로 폼을 제출할 수 있게 해줍니다. 또한 이는 HTML 표준에 맞는 방식으로, 스크린 리더와 같은 보조 기술에서도 폼 제출 버튼을 명확히 인식할 수 있도록 도와줍니다. 

따라서, button type="submit"을 사용하여 폼 제출 동작을 표준에 맞게 처리하는 것이 사용성과 접근성 측면에서 적합합니다. 

```ts
<form onSubmit={handleSubmit>
    <button type='submit'>제출</button>
</form>
```

아래 글도 한번 읽어보시면 도움이 되실거예요

https://stackoverflow.com/questions/6908187/form-onsubmit-versus-submit-button-onclick
https://tecoble.techcourse.co.kr/post/2021-05-22-html-form-tag/

hanheel replied:
step1에서 div태그를 form 태그로 바꾸면서 onSubmit 이벤트를 사용하려고 했는데 완전히 까먹고 있었네요 🥲
문서를 읽어보니 확실히 어떤 부분에서 필요한지도 파악하게 된 것 같습니다! 
습관적으로 onClick 이벤트를 사용하게 되는 것 같은데 꼭 신경써서 고쳐보도록 하곘습니다 감사합니다 :)

✅ 커밋 : [415707c](https://github.com/woowacourse/react-payments/pull/465/commits/415707c97841bb15f52df5ef074764f8cabc9e9d)

wonsss commented on src/component/card/CardNumber.tsx:
스토리북이 있는 컴포넌트도 있고 없는 컴포넌트들도 있던데, 어떤 기준으로 컴포넌트별 스토리북 작성 여부를 결정하셨을까요?

hanheel replied:
아래와 같은 컴포넌트는 따로 스토리북을 작성하지 않았습니다!

**1. 상태변화가 없거나, 하나의 조합으로만 사용되는 컴포넌트**
ex) `Description` , `Button`
* 현재 Description 컴포넌트는 CardDescriptiveSelect 내부에서만 사용되고, 다른 곳에서 독립적으로 재사용되지 않으며, 항상 같은 텍스트 조합으로만 쓰입니다. 또한, 상태에 따라 변하지 않습니다! 

**2. 조금 더 큰 단위, 또는 작은 단위로 포괄해서 테스트 하는게 용이한 경우**
ex) `CardDescriptiveSelect` , `CardSelect`, `FormField` , `CardDescriptiveInput` , `CardInput`
* CardInput과 CardSelect는 주로 에러 발생 시 스타일 변화가 핵심 시나리오이기 때문에, 이 부분은 CardLabeledInput 등의 상위 Story에서 충분히 테스트되고 있어 따로 분리하지 않았습니다.
* CardDescriptiveInput 계열은 LabeledInput + Description의 조합이기 때문에, 상태 변화는 전부 LabeledInput 측에서 다루고 있어, Description 단독의 Story는 내용 중복만 발생한다고 판단했습니다.

저는 스토리북이 상태 변화에 따른 (시나리오에 따른) 컴포넌트의 모습을 잘 보여준다고 생각해서 이런 기준을 갖고 스토리북을 작성할지 안할지를 결정했는데, **협업** 측면에서는 제작한 컴포넌트 대부분을 시각화 하는 것이 좋을까요??

wonsss commented on src/component/card/CardNumber.tsx:
```diff
@@ -0,0 +1,51 @@
+        <CardInformation>{cardInput.first}</CardInformation>
+        <CardInformation>{cardInput.second}</CardInformation>
+        <CardMaskingInformation>
+          {cardInput.third && maskingNumber(String(cardInput.third).length)}
+        </CardMaskingInformation>
+        <CardMaskingInformation>
+          {cardInput.fourth && maskingNumber(String(cardInput.fourth).length)}
+        </CardMaskingInformation>
+      </CardNumberContainer>
+
+      <CardInformation>
+        {cardInput.MM && `${cardInput.MM}`}
+        {cardInput.YY && `/${cardInput.YY}`}
```
`cardInput.MM && `${cardInput.MM}`` 이런 식의 && 연산자를 사용한 조건문은 단락평가를 사용하여 첫 번째 값이 falsy 일 경우 두 번째 값을 평가하지 않고 바로 그 값을 반환합니다.
현재 첫 번째 값으로 둔 타입이 아마 string이라서 string의 falsy 값인 `""` 이 들어가도 `""`  만 반환하여 화면에 아무것도 노출되지 않으므로 문제는 없을 것 같아요.
다만, 다른 falsy인 `0`,`NaN`,`false` 같은 값들이 첫번째 값에서 평가된다면 이들이 반환되어 화면에 표시될 수 있습니다. 이게 의도와 다른 구현일 가능성이 있어, `&&`로 조건부 렌더링할 때는 첫번째 값이 Boolean 타입이 확실할 때 사용하는 것이 좋습니다. 또는 `?` 삼항조건연산자를 사용하여 명시적으로 조건부 렌더링하는 것도 좋아요.

아래 아티클 내용도 참고해보세요~
https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx

hanheel replied:
처음에는 input은 무조건 문자열을 반환하니까 위와 같이 && 연산자를 사용해도 문제가 발생하지 않을 거라 생각했는데, 프로젝트가 확장되면 이런 사소한 부분에서 문제가 발생할 수도 있겠네요..! 명시적 조건부 렌더링에 대해 이제 명확히 이해한 것 같습니다!! 감사합니다.

✅ 커밋 : [0e74f04](https://github.com/woowacourse/react-payments/pull/465/commits/0e74f04f43ce4f60775342569eb7d692c783ddcf)

wonsss commented on src/component/input/CardDescriptiveInput.tsx:
```diff
@@ -0,0 +1,51 @@
+import { FormSectionContainer } from '../../styles/FormSectionStyle';
+
+interface ValueProps {
+  [key: string]: string | string;
```
`string | string`은  실수로 동일한 타입을 중복으로 적으신 것 같아요. string으로 단순하게 수정해주세요~!

hanheel replied:
반영했습니다! 

✅ 커밋 : [1970980](https://github.com/woowacourse/react-payments/pull/465/commits/1970980e8e19df1dda9c4b4108a0056607364d4f)

wonsss commented on src/component/input/CardLabeledInput.tsx:
```diff
@@ -0,0 +1,47 @@
+  value: ValueProps;
+}
+
+const CardLabeledInput = ({ config, errorMessage, handleCardInput, isErrors, value }: CardInputGroupProps) => {
```
CardLabeledInput 보다 더 적절한 컴포넌트 명을 고려해보시면 좋을 것 같아요.
현재 컴포넌트의 역할을 보면 여러 개의 CardInput을 렌더링하고 관리하고 있어요.
한 개의 input이 아니기 때문에, 그룹이라는 의미를 이름에서 주는 게 가독성에 좋을 것 같습니다.
따라서 그룹 의미를 담아 CardInputGroup을 추천드려요.

이름은 다른 개발자들이 봤을 때도 직관적으로 이해할 수 있을지 고민하며 짓는 연습을 하시면 좋습니다.

hanheel replied:
네이밍 지을 때마다 읽는 입장이 돼서 생각해 보겠습니다!

wonsss commented on src/component/input/CardDescriptiveInput.tsx:
```diff
@@ -0,0 +1,51 @@
+  handleCardInput: (inputKey: keyof CardInputProps, value: string) => void;
+}
+
+const CardDescriptiveInput = ({
```
CardDescriptiveInput 이름 또한 개선할 여지가 있을 것 같아요.
이름만 보고서는 `Card`와 관련된 `서술적인` `Input` 한 개가 있는 컴포넌트인가? 라는 생각이 드는데, 실제 역할과 다른 것 같아요.

hanheel replied:
`CardFormSection` 이라는 네이밍이
Input 여러개 + label + Description 을 포함한 form의 한 덩어리라는 의미 +  `CardInputGroup`라는 네이밍과 계층적으로도 조화가 되는 것 같아 으로 변경해보았습니다. 

(네이밍이 쉽지 않네요..😂)

✅ 커밋 : [ee1926a](https://github.com/woowacourse/react-payments/pull/465/commits/ee1926afcf84f9c792ff980cecf9a798734c1eda)

wonsss commented on src/component/input/CardDescriptiveInput.tsx:
```diff
@@ -0,0 +1,51 @@
+  handleCardInput,
+}: CardDescriptiveInputProps) => {
+  return (
+    <FormSectionContainer>
+      <Description headText={descriptionText.headText} detailText={descriptionText.detailText} />
+      <CardLabeledInput
+        config={config}
+        value={value}
+        errorMessage={errorMessage}
+        handleCardInput={handleCardInput}
+        isErrors={isErrors}
+      />
+    </FormSectionContainer>
```
> 컴포넌트의 계층 구조
혹시 현재 코드 구조에서 복잡성을 더 줄이기 위해 추가로 시도해볼 수 있는 방법이 있을지 궁금합니다.
(의도 2번) config를 통해 일부의 props를 전달하는 방식이 코드를 읽을 때 바로 props를 파악하기 어렵게 만들거나 컴포넌트 간 연결 관계를 추적하는데에 오히려 복잡성을 증가시킨건 아닌가 하는 고민이 듭니다. 현재 수준의 웹에서 위와같은 방식 외에 시도해 볼 수 있는 다른 방식이 있을까요? (config를 사용한 방식 외에, 복잡성을 줄이기 위해 제가 시도해 볼 수 있는 방식이 있을까요?)

관심사 분리 측면에서 컴포넌트 분리를 잘 해주셨다고 생각해요.
더 개선을 해본다면, 컴포넌트를 조립한다는 관점으로 개발해보면 좋을 것 같아요.
CardDescriptiveInput 컴포넌트는 자신이 사용하지 않고 props drilling하기 위해 받는 props가 많습니다.
그렇다면, CardDescriptiveInput 내에서 `<CardLabeledInput />` 부분을 `{children}` 으로 뚫어 놓고 사용처에서 받아 조립하면 어떨까요? 

```tsx
const CardDescriptiveInput = ({ descriptionText, children }: CardDescriptiveInputProps) => {
  return (
    <FormSectionContainer>
      <Description headText={descriptionText.headText} detailText={descriptionText.detailText} />
      {children}
    </FormSectionContainer>
  );
```

그리고 CardDescriptiveInput 컴포넌트 명도 폼의 한 섹션을 구성한다는 의미 정도로 FormSection 같은 것으로 수정해도 괜찮을 것 같아요.

아무튼 이러한 조합 관점을 더 확장하면, Compound Component Pattern 이라는 것이 있는데, 이것도 더 알아보시면 도움이 되실 거예요.
https://patterns-dev-kr.github.io/design-patterns/compound-pattern/

wonsss reviewed (changes requested):
안녕하세요! 블루 😄 
2단계 고생 많으셨습니다! 
커스텀 훅 사용을 잘 해주셨고 컴포넌트 재사용도 잘 고민해주셨네요! 👍 
UX도 많이 고민해주신 것 같아 좋았습니다.
질문에 대한 답변은 관련 코드 맥락에 달아두었습니다.
리뷰 반영 후 재리뷰 요청주세요! 감사합니다 🙇

hanheel committed:
refactor : 사용하지 않는 파일 삭제

hanheel committed:
refactor : pages 컴포넌트 --Page로 네이밍 변경

hanheel committed:
refactor : return문에 중괄호를 사용하도록 통일

hanheel commented:
/c-rabbit ignore

hanheel committed:
refactor : PATH를 상수로 관리

hanheel committed:
refactor : isNumber 함수명과 리턴값 통일

hanheel committed:
refactor : isError 와 errorMessage 상태 통합하기

hanheel committed:
refactor : onClick 이벤트를 onSubmit 이벤트로 수정

hanheel committed:
refactor : 동일 타입 중복 삭제

hanheel committed:
refactor : input, select 컴포넌트 네이밍 변경

1. CardLabeledInput -> CardInputGroup
2. CardDescriptiveInput -> CardFormSectionInput

hanheel committed:
refactor : 명시적으로 조건부 렌더링

hanheel commented:
안녕하세요 마르코!
이번 주에는 페어와 함께 새로운 미션을 진행하느라 리뷰 반영이 다소 늦어졌습니다 🥲

공유해주신 아티클들 모두 감사드립니다! 하나씩 읽어보면서 어렴풋하게 알고 있던 개념들을 정리할 수 있어서 많이 도움이 되었습니다 🙏🏻

리뷰 내용은 커밋을 통해 반영하였고, 궁금한 부분은 각 코멘트에 답글로 남겨두었습니다.
확인 부탁드리겠습니다! 👍🏻

wonsss reviewed (approved):
안녕하세요! 블루 😄 
리뷰 반영 잘해주셔서 감사합니다. 👍 
리뷰 드린 코멘트에 블루도 자세히 답변을 주셔서 정확히 어떤 점을 이해하셨고 또는 고민이신지 알 수 있어서 좋은 피드백 과정이었다고 생각해요!
2단계 목표에 맞게 코드는 충분히 잘 수정된 것 같아 여기서 페이먼츠 미션은 마무리해도 될 것 같아요.
수고 많으셨습니다! 화이팅입니다~! 💪

hanheel commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.
